### ISSUE TITLE
Can't pull docker images containing device nodes when running docker inside user namespace

### ISSUE THREAD

**Issue by willbryant** [moby/moby#35245]
**Description**

The Docker CE daemon mostly works inside a user namespace, allowing it to be used in environments such as virtual servers implemented using LXD.

However, using most graphdrivers, it is not possible to pull any image that has device nodes, such as the `ubuntu` images, because a "docker: failed to register layer: Error processing tar file(exit status 1): operation not permitted." error occurs.

**Steps to reproduce the issue:**
1. Start a user namespace, eg. by starting an LXD container (giving it the security.nesting permission)
2. Install Docker CE from the normal repository.  Pull and run the hello-world program to confirm that it can run some containers.
3. Try to pull an image containing device nodes eg. `docker pull ubuntu:latest`.

**Describe the results you received:**

A permission denied error occurs while untaring one of the image layers:

```
# docker pull ubuntu:latest
latest: Pulling from library/ubuntu
ae79f2514705: Extracting [==================================================>]  47.62MB/47.62MB
5ad56d5fc149: Download complete 
170e558760e8: Download complete 
395460e233f5: Download complete 
6f01dc62e444: Download complete 
failed to register layer: Error processing tar file(exit status 1): operation not permitted
```

**Describe the results you expected:**

The image should pull successfully, and the device nodes should be ignored (as they will be ignored at runtime anyway).

**Additional information you deem important (e.g. issue happens only occasionally):**

This bug was mostly fixed in 617c352, but it looks like one case was missed.

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-97-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.32GiB
Name: test-nwtn325
ID: PA6Z:7GWM:WVIF:JUPM:TAEM:AONA:VBOA:47JY:4OIJ:LAZZ:3533:Z7W2
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

LXD 2.0, or any other environment that uses user namespaces.

**Comment by willbryant** [comment]
Did some more testing; I found that overlay2 has this problem, but overlay doesn't.

**Comment by estesp** [comment]
Probably need to go back and look at Serge's patches originally to fix this. Didn't realize they were graph driver specific but maybe they were.

**Comment by willbryant** [comment]
Looks to me like the main fix was 617c352?  It patched daemon/graphdriver/overlay/copy.go, but nothing in daemon/graphdriver/overlay2.

What confuses me is that commit, @estesp added a special case to the os.ModeSocket case in  overlay/copy.go, but not the case os.ModeDevice case.  But it works anyway!  At least, well enough that I can pull an ubuntu image.

**Comment by willbryant** [comment]
Could it be the special case in handleTarTypeBlockCharFifo?  But then, why wouldn't it work for overlay2 as well?

**Comment by hallyn** [comment]
No, the copyDir is called on a directory, so so it would definitely trip over that mknod.  Can't say I remember what the deal was there, and I do see the files are completely gone now :)

**Comment by remingtonc** [comment]
Any update on this issue? Can confirm that switching to ```overlay``` instead of ```overlay2``` resolves this problem. However it appears [strongly discouraged](https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/#configure-docker-with-the-overlay-or-overlay2-storage-driver) per the docs. Will using overlay instead of overlay2 create any legitimate issues?

We set up a development environment with physical machines partitioned in to LXCs per developer. Not being able to run Docker to its fullest really detracts from the value add of that partitioning.

**Comment by estesp** [comment]
@remingtonc I found the issue, I believe. I'll pull a proper PR together to get `overlay2` back in line w.r.t. running in a user namespace.

**Comment by estesp** [comment]
PR #35794 opened to fix this issue

**Comment by remingtonc** [comment]
@estesp Amazing turnaround, thank you!

**Comment by willbryant** [comment]
I think this must have not made it into 18.03 CE, as I can't see it in the release notes and the problem persists :(

**Comment by thaJeztah** [comment]
hm; the fix should be in 18.01 and up; wonder if there's more to fix. Let me reopen 

ping @estesp

**Comment by estesp** [comment]
Let me see if I can get my test env. up with 18.03 and reproduce; very strange

**Comment by remingtonc** [comment]
This is also being semi-tracked [on StackOverflow](https://stackoverflow.com/questions/45731683/docker-pull-operation-not-permitted/47800929#47800929), [this answer](https://stackoverflow.com/a/45934747/1357467) mentions it is by design - is this true? Or should this not be an issue?

Sidebar: Are moby releases coordinated with Docker releases, or does Docker cherrypick from moby independently?

**Comment by estesp** [comment]
Well, I'm definitely confused now. There is no issue with device nodes after the PRs that have been merged, however, the `EPERM` is coming from [this code](https://github.com/moby/moby/blob/master/pkg/archive/archive_linux.go#L70) that is handling overlay whiteout translation in pkg/archive/archive_linux.go. This code is only used for overlay2 and not overlay, hence the reason overlay is working and not overlay2.

I see that in the kernel overlay driver, this has been handled for years with patches that "override" the remapped container root with privilege to do these operations (see [this patchset](http://ubuntu.5.x6.nabble.com/PATCH-RFC-overlayfs-xattr-allow-unprivileged-users-to-whiteout-td5058630.html) from @hallyn back in 2014) if I understand correctly.

Apparently when I tested I didn't try pulling any images that had whiteout entries (e.g. like a distro image with no deletions in upper layers), hence why I didn't find the issue. `docker pull redis` is an easy testcase as it isn't a huge image and the 4th or 5th layer has whiteout files that reproduce the issue.

Any thoughts @dmcgowan?

**Comment by justincormack** [comment]
We really ought to remove the device nodes from the Ubuntu images anyway - they are pointless.

**Comment by willbryant** [comment]
I'm definitely no expert, but doing that looked non-trivial to me because they are in the base tarball, not one of the later layers.  And that base tarball is shared with the VM cloud images, I think?

**Comment by willbryant** [comment]
@estesp is it device nodes that being "whited out" that's causing this remaining problem, or any files?

**Comment by estesp** [comment]
@willbryant no; device nodes are no longer the issue here; it is the overlay2 whiteout implementation that is being denied the `xsetattr` permission.

The following simple `Dockerfile` can reproduce the issue (I've pushed the image as **estesp/overlay2test:whiteout** if anyone wants to easily see the issue):
```
FROM alpine:latest

WORKDIR /
RUN mkdir somedir
COPY somefile /somedir/
RUN echo "hi" >test.file
RUN rm -fR /somedir/somefile
```

The layer created by the final `RUN` command includes a whiteout entry that, during `docker pull` to a userns-enabled daemon will fail with `EPERM`; only when using the `overlay2` storage driver.

**Comment by thaJeztah** [comment]
ping @tianon ^^ 🤗

**Comment by AkihiroSuda** [comment]
Here is a patch for this issue: https://github.com/rootless-containers/usernetes/tree/master/build/patches/moby

(I'd like to open PR soon but still waiting for https://github.com/opencontainers/runc/pull/1862 to be merged...)

**Comment by AkihiroSuda** [comment]
binary (Aug 28) is available here: https://github.com/rootless-containers/usernetes/releases

Will rebase the patch next week

**Comment by AkihiroSuda** [comment]
after all, I rebased the patch today: https://github.com/rootless-containers/usernetes/pull/38

`docker run -it --rm ubuntu` and `docker run -it --rm centos` with overlay2 works for me

**Comment by dhiltonp** [comment]
Thanks, @AkihiroSuda!

I tested your patch and it wasn't sufficient. 

I've opened #37966 to fix my issue - any input is welcome!
